### PR TITLE
Fix bodyswap to work on the latest adv-beta

### PR DESCRIPTION
**Pull request by Crystalwarrior**
Also adds automatic re-centering on the unit you bodyswapped to!

### CODE REVIEW

File: bodyswap.lua [hunk 1]
```diff
@@ -30,15 +30,16 @@ function clearNemesisFromLinkedSites(nem)
...
+            if site.populace.nemesis[i] == nem.id then
+                site.populace.nemesis:erase(i)
             end
         end
     end
 end
 
 function createNemesis(unit)
+    -- TODO: figure out how to create a nemesis entry without relying on external scripts
     local nemesis = reqscript('modtools/create-unit').createNemesis(unit, unit.civ_id)
```
Comment (myk002) [inline bodyswap.lua:43]: `local nemesis = unit:create_nemesis(false, true)`

File: bodyswap.lua [hunk 2]
```diff
@@ -149,9 +126,11 @@ function swapAdvUnit(newUnit)
     setNewAdvNemFlags(newNem)
     configureAdvParty(newNem)
     df.global.adventure.player_id = newNem.id
-    activeUnits[newUnitIndex] = oldUnit
-    activeUnits[oldUnitIndex] = newUnit
+    df.global.world.units.adv_unit = newUnit
     oldUnit.idle_area:assign(oldUnit.pos)
+
+    -- +1 is necessary because in adv mode, the window pos is off by 1 from dwarf mode for some reason
```
Comment (myk002) [inline bodyswap.lua:132]: if this is systematically true, we'll need to fix it in Gui.cpp, not here. Could you test further to see if we're also off by 1 in fort mode?

File: bodyswap.lua [hunk 3]
```diff
@@ -149,9 +126,11 @@ function swapAdvUnit(newUnit)
...
+    dfhack.gui.revealInDwarfmodeMap(xyz2pos(newUnit.pos.x+1, newUnit.pos.y+1, newUnit.pos.z), true)
 end
 
 if not dfhack_flags.module then
```
Comment (myk002) [inline bodyswap.lua:136]: the help command (in the print statement below) should be `help bodyswap`, not `bodyswap -help`.

**Review by myk002 — COMMENTED** [review]
also needs a line in the changelog. New Tools, "reinstated" format.

File: bodyswap.lua (hunk 2 above)
Comment (myk002) [reply bodyswap.lua:132]: as per https://github.com/DFHack/dfhack/pull/4660 it looks like we need to fix this in `revealInDwarfmodeMap` itself. Would you like to submit that fix?

File: bodyswap.lua [hunk 4]
```diff
@@ -30,16 +30,16 @@ function clearNemesisFromLinkedSites(nem)
...
     for _, link in ipairs(nem.figure.site_links) do
         local site = df.world_site.find(link.site)
-        for i = #site.unk_1.nemesis - 1, 0, -1 do
-            if site.unk_1.nemesis[i] == nem.id then
-                site.unk_1.nemesis:erase(i)
+        for i = #site.populace.nemesis - 1, 0, -1 do
+            if site.populace.nemesis[i] == nem.id then
+                site.populace.nemesis:erase(i)
             end
         end
```
Comment (myk002) [inline bodyswap.lua:37]:
```suggestion
        utils.erase_sorted(site.populace.nemesis, nem.id)
```

File: bodyswap.lua [hunk 5]
```diff
@@ -149,9 +125,10 @@ function swapAdvUnit(newUnit)
     setNewAdvNemFlags(newNem)
     configureAdvParty(newNem)
     df.global.adventure.player_id = newNem.id
-    activeUnits[newUnitIndex] = oldUnit
-    activeUnits[oldUnitIndex] = newUnit
+    df.global.world.units.adv_unit = newUnit
     oldUnit.idle_area:assign(oldUnit.pos)
+
+    dfhack.gui.revealInDwarfmodeMap(newUnit.pos, true)
```
Comment (myk002) [inline bodyswap.lua:131]:
```suggestion
    dfhack.gui.revealInDwarfmodeMap(dfhack.units.getPosition(newUnit), true)
```

File: bodyswap.lua (hunk 4 above)
Comment (myk002) [reply bodyswap.lua:37]: `populace.nemesis` is a sorted vector

File: bodyswap.lua (hunk 5 above)
Comment (myk002) [reply bodyswap.lua:131]: in case you've bodyswapped into a caged unit

**Review by myk002 — COMMENTED** [review]
needs "reinstated" changelog entry

**Comment by myk002** [comment]
As rome of oxtrot mentioned on [Discord](https://discord.com/channels/793331351645323264/807443277132595211/1246696330672209931), please also add your name to [Authors.rst](https://github.com/DFHack/dfhack/blob/develop/docs/about/Authors.rst)

**Comment by myk002** [comment]
code looks reasonable, but from Discord I'm gathering there are still open questions about the behavior of this script, is that right?

btw still needs `changelog.txt` entry

File: bodyswap.lua
Comment (myk002) [inline bodyswap.lua:1]: nem.flags.RETIRED_ADVENTURER is no longer a flag -- it was unused and has been removed. you no longer need to set it to false on line 17

**Review by myk002 — COMMENTED** [review]
needs changelog entry

**Comment by Crystalwarrior** [comment]
Since this fixes the functionality the script came with originally, step 2 is making the script not auto add to core party and resolve any retire issues on adventurers that become wanderers post body swap and look into how the game handles those units

**Review by myk002 — APPROVED** [review]
Could you create an issue on GitHub detailing the surrounding issues you've found with the script's behavior? I want to make sure they're recorded so we don't forget about them.

**Pull request merged**